VEP#148: Persistent MAC Addresses for VirtualMachines

RamLavi opened the pull request:
### VEP Metadata
**Tracking issue**: https://github.com/kubevirt/enhancements/issues/148
**SIG label**: /sig network

### What this PR does
Introduce a new VEP proposing automatic MAC address persistence for VirtualMachines by adding controller logic to virt-controller that persists runtime MAC addresses from VMI status back to VM template spec.

### Special notes for your reviewer

RamLavi committed:
VEP#148: Persistent MAC Addresses for VirtualMachines

Introduce a new VEP proposing automatic MAC address persistence for
VirtualMachines by expanding network synchronizer logic on virt-controller to
persists runtime MAC addresses from VMI status back to the VM template spec
and VMI spec.

Signed-off-by: Ram Lavi <ralavi@redhat.com>

orelmisan commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,324 @@
+## Goals
+
+- Automatically persist MAC addresses from VMI status to VM spec without user intervention
+- Support all interface types (masquerade, bridge, SR-IOV, custom bindings like l2bridge) that showcase MAC-Address on VMI.status.interfaces
```
`l2bridge` is a downstream term. The upstream term is `managedTap`.

orelmisan commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,324 @@
+
+**As an application developer**, I want VMs to behave like physical machines where MAC addresses don't change on reboot, so my applications that depend on network identity work correctly.
+
+**As a VM operator**, I want to migrate a VM across namespaces (e.g., in hosted control plane scenarios) while preserving MAC addresses, so the guest OS doesn't detect a hardware change.
```
nit: Please consider grouping the user stories by persona.

RamLavi replied:
DONE

orelmisan reviewed (changes requested):
Thank you for the PR @RamLavi!
Please see the inline comments.

dasionov reviewed (commented):
Hi @RamLavi, I understand the design and overall goals. In addition to @orelmisan’s comments, I’d like to add that this approach feels very similar to how we handled the firmware UUID persistence, and to me it seems more natural to apply the same pattern here.

If it’s feasible, perhaps this logic could be integrated into the existing network synchronizer, or alternatively implemented as a new dedicated synchronizer for this purpose.

Introducing an entirely new controller solely for patching a field feels a bit heavy-handed in my opinion.

```
	if c.netSynchronizer != nil {
		syncedVM, err := c.netSynchronizer.Sync(vmCopy, vmi)
		if err != nil {
			return vm, vmi, handleSynchronizerErr(err), nil
		}
		vmCopy.ObjectMeta = syncedVM.ObjectMeta
		vmCopy.Spec = syncedVM.Spec
	}

	if c.firmwareSynchronizer != nil {
		syncedVM, err := c.firmwareSynchronizer.Sync(vmCopy, vmi)
		if err != nil {
			return vm, vmi, handleSynchronizerErr(err), nil
		}
		vmCopy.ObjectMeta = syncedVM.ObjectMeta
		vmCopy.Spec = syncedVM.Spec
	}
```

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,324 @@
+When a VM is created without explicit MAC addresses specified in its spec:
+1. VM is created with empty `spec.template.spec.domain.devices.interfaces[].macAddress` fields
+2. VMI is instantiated from the VM template
+3. CNI/network plugin assigns MAC addresses at pod creation time
```
OK, but regardless - as long as it lands on status.interfaces then we can persist it to the VM.

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,324 @@
+### Real-World Impact
+
+Kubevirt opinionated deployment ([hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)) currently work around this by:
+- relying on external MAC management tools (kubemacpool) which introduce webhook complexity
```
DONE

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
````diff
@@ -0,0 +1,324 @@
+Add a VM controller that will persist runtime MACs:
+
+```go
+func (c *VMController) persistRuntimeMACsToVMSpec(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
````
Changed to explanation. moved to code to a sub-section (we can move it to the end of the doc if you think it best)

RamLavi commented:
> Hi @RamLavi, I understand the design and overall goals. In addition to @orelmisan’s comments, I’d like to add that this approach feels very similar to how we handled the firmware UUID persistence, and to me it seems more natural to apply the same pattern here.
> 
> If it’s feasible, perhaps this logic could be integrated into the existing network synchronizer, or alternatively implemented as a new dedicated synchronizer for this purpose.
> 
> Introducing an entirely new controller solely for patching a field feels a bit heavy-handed in my opinion.
> 
> ```
> 	if c.netSynchronizer != nil {
> 		syncedVM, err := c.netSynchronizer.Sync(vmCopy, vmi)
> 		if err != nil {
> 			return vm, vmi, handleSynchronizerErr(err), nil
> 		}
> 		vmCopy.ObjectMeta = syncedVM.ObjectMeta
> 		vmCopy.Spec = syncedVM.Spec
> 	}
> 
> 	if c.firmwareSynchronizer != nil {
> 		syncedVM, err := c.firmwareSynchronizer.Sync(vmCopy, vmi)
> 		if err != nil {
> 			return vm, vmi, handleSynchronizerErr(err), nil
> 		}
> 		vmCopy.ObjectMeta = syncedVM.ObjectMeta
> 		vmCopy.Spec = syncedVM.Spec
> 	}
> ```

@dasionov great idea! I've incorporated this to the VEP. PTAL

EdDev commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,324 @@
+1. VM is created with empty `spec.template.spec.domain.devices.interfaces[].macAddress` fields
+2. VMI is instantiated from the VM template
+3. CNI/network plugin assigns MAC addresses at pod creation time
+4. MAC addresses appear in VMI status but **not** in VM spec
```
@orelmisan , the text is pretty general, how will this detail matter to the understanding of the current state?

RamLavi replied:
yes you are right. 
I've re-added the SRI-IOV to the goals, and specified the proper handling in the `edge cases handled` section. 
PTAL.

EdDev commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,324 @@
+
+- Removal of Cluster-wide MAC address allocation feature from kubemacpool (should happen in tandem, but is not part of this VEP)
+- Cluster-wide MAC address collision detection (handled by external tools like kubemacpool)
+- Persistence for pod-based workloads (only VM/VMI)
```
@RamLavi , please be explicit and add it to the text please.

EdDev replied:
@orelmisan , can this be resolved now?

EdDev reviewed (commented):
Thank you!!

First pass, up until the `Design` section (not including).

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,338 @@
+2. VMI is instantiated from the VM template
+3. CNI/network plugin assigns MAC addresses at pod creation time
+4. MAC addresses appear in VMI status but **not** in VM spec
+5. On VM restart/stop-start, **new MAC addresses are assigned** (VM spec is still empty)
```
Added

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,338 @@
+## Goals
+
+- Automatically persist MAC addresses from VMI status to VM spec without user intervention
+- Support all interface types (masquerade, bridge, SR-IOV, custom bindings like managedTap) that showcase MAC-Address on VMI.status.interfaces
```
DONE

RamLavi commented:
[Change](https://github.com/kubevirt/enhancements/compare/7fc43e0f0ed0ab46ed43cc53ffcab2e0f28217cc..9a8c5a2cfe7c23cb891b79b27a12a9fa6bfa10ac): Address @EdDev 's initial review

EdDev commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+### Events
+
+Successful MAC persistence will generate a Kubernetes Event on the VM:
```
Is this something that exists with the firmware UID? If not, why do we need it?

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,339 @@
+### Performance Impact
+
+- **Controller overhead**: Minimal - synchronizer runs as part of existing VM reconciliation loop
+- **API server load**: One additional PATCH per VM on first start (subsequent starts are no-op)
```
added this as a commentary, but I don't think it's really needed. This section is only there to show that the performance impact is minimal-moderate.
PTAL

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+## Functional Testing Approach
+
+### Unit Tests
```
added  a unit test case for this.

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,339 @@
+- VM cloning workflow: Clone VM → Verify MAC preserved in clone
+- Cross-namespace scenario: Create VM in ns-a with MAC, recreate in ns-b with same spec → MAC preserved
+
+## Implementation Phases
```
Added graduation criteria. I'm aiming to start with beta, pending community's approval

RamLavi commented:
[Change](https://github.com/kubevirt/enhancements/compare/9a8c5a2cfe7c23cb891b79b27a12a9fa6bfa10ac..0cfa471fb92500376bcc0bc4ffa7a95dd4d2d41f): 
- addressed @EdDev 's review
- added Graduation criteria

nirdothan reviewed (commented):
Thanks @RamLavi! Looks good.  See my few comments below.

nirdothan commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,286 @@
+- Removal of Cluster-wide MAC address allocation feature from kubemacpool (should happen in tandem, but is not part of this VEP)
+- Cluster-wide MAC address collision detection (handled by external tools like kubemacpool)
+- MAC Persistence for ephemeral entities such as pods, VMIs (except those managed by a VM)
+- MAC Persistence for SR-IOV interfaces (SR-IOV status MAC is only populated when explicitly set in VM spec)
```
That's quite unfortunate because we currently [don't even persist guest interface names across live migrations](https://issues.redhat.com/browse/CNV-74876), so this could be where it's needed the most.

RamLavi replied:
I understand. I've added the SR-IOV case back into the goals. I think we can support this with a small primer PR (that will update SRIOV interfaces using the multus-status). Updated the VEP.

RamLavi commented:
[Change](https://github.com/kubevirt/enhancements/compare/0cfa471fb92500376bcc0bc4ffa7a95dd4d2d41f..da138d44cf518079486f090f249d66f9a9d0a3d2): addressed the concerns pointed out by @orelmisan @nirdothan - PTAL

nirdothan reviewed (approved):
Thank you!

orelmisan commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,313 @@
+
+**As a cluster admin**, I want to clone a VM to another namespace and preserve its MAC addresses, so that the guest OS environment remains consistent without manual MAC copying.
+
+**As an application developer**, I want VMs to behave like physical machines where MAC addresses don't change on reboot, so my applications that depend on network identity work correctly.
```
This user story seems to have some overlap with the first one. Could you clarify the key differences or unique requirements for this specific case?

RamLavi replied:
I agree with @EdDev, it has merit IMO. But if more people say that this is redundant then I won't insist keeping it

EdDev commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,313 @@
+**Active Migrations:**
+- Synchronizer skips persisting MACs when VMI is an active migration source 
+- This prevents persisting the source VMI's MAC when it's about to be replaced by the target VMI
+- After migration completes, the target VMI's MAC is persisted (ensuring the correct, surviving MAC is preserved)
```
This is mainly describing a cross-namespace (or cluster) migration, which is a special case.
In a regular live migration, there is only one single VMI.

 I guess this can be generalized so the synchronizer will just avoid the reflection of the mac during any live migration, which should in practice never happen, because the mac is reflected on the status once the domain is created and that is expected to happen before or in parallel to the domain being detected as "running".
 
 I would prefer this would be just a backup safety measurement, but I am also fine to keep it.

EdDev commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,313 @@
+
+**SR-IOV MAC Persistence:**
+Currently, SR-IOV status MAC is only [populated](https://github.com/kubevirt/kubevirt/blob/59c62130ead7e3b05b3de737f71aba3aae58e798/pkg/network/setup/netstat.go#L307) when explicitly set in VM spec.
+So in case of SR-IOV, a prerequisite step would take place to extract the MAC address from the pod's `network-status` annotation into `vmi.Status.Interfaces[].MAC`, following the existing pattern for IPs and DeviceInfo extraction. The MAC synchronizer will then persist it to VM spec as normal.
```
+1 to be accurate on what is put on writing here.
But please do not enter too low level details, like the virt-handler having to consider things in its logic. It feels too low level to me.

EdDev reviewed (commented):
Thank you @RamLavi , this looks really good.
The active comments are pretty minor, hope we can wrap this one up quickly (while it is fresh in my mind).

frenzyfriday commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,313 @@
+
+## Non Goals
+
+- Removal of Cluster-wide MAC address allocation feature from kubemacpool (should happen in tandem, but is not part of this VEP)
```
So after this VEP will GA tools like kubemacpool would have to be uninstalled by the cluster admins? If so should KubeVirt put a warning during upgrade?

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,313 @@
+This creates several operational problems:
+
+**Guest OS Networking Instability**
+- Some Guest OS networking configurations are  bound to MAC addresses (DHCP leases, network bonds, udev rules, firewall configurations)
```
DONE

RamLavi commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,313 @@
+- Idempotent design allows safe reconciliation after controller restart
+
+**Active Migrations:**
+- Synchronizer skips persisting MACs when VMI is an active migration source 
```
>Is this referring to a cross{cluster, namespace} live migration?

yes.

>What would happen on regular migrations?

As Eddy said - in regular live migration the VMI does not get duplicated, only the virt-launcher pod.



@EdDev - 
+1
My concern is e2e tests that will start migration right after the VMI is running.

RamLavi commented:
[Change](https://github.com/kubevirt/enhancements/compare/08e3e5bad63f2ca3a6bf8fd5413edc00a748c490..6e1bf6274b65910ad00f40dd0de5e059981c4c88): Address @EdDev @orelmisan offline comments.

EdDev reviewed (approved):
Thank you, this looks great!

orelmisan commented on veps/sig-network/persistent-mac-addresses.md:
```diff
@@ -0,0 +1,312 @@
+
+### VM Export Race Condition
+
+The VM export feature has a similar timing issue as migration. If a VM is exported before the MAC synchronizer has persisted the MAC addresses to `vm.Spec`, the exported VM snapshot will not contain the runtime MAC addresses. However, unlike migration (which has observable VMI status indicating an active migration), VM export does not expose any status on the VM object that the synchronizer can d ...
```
nit: s/MAC synchronizer/network synchronizer

orelmisan reviewed (approved):
Thank you @RamLavi !

oshoval commented:
Please consider https://github.com/kubevirt/kubevirt/issues/16422#issuecomment-3701656440
i.e what happens if someone changes the MAC in the guest (MAC will be changed in status).
if relevant, it also mentions a possible solution, i.e report the permanent HW address.
We can also just report taint message in this case as libvirt does if custom guest agent change the guest.

EdDev commented:
> Please consider [kubevirt/kubevirt#16422 (comment)](https://github.com/kubevirt/kubevirt/issues/16422#issuecomment-3701656440) i.e what happens if someone changes the MAC in the guest (MAC will be changed in status). if relevant, it also mentions a possible solution, i.e report the permanent HW address. We can also just report taint message in this case as libvirt does if custom guest agent change the guest.

If someone changes the mac in the guest, it will appear as unbounded, i.e. Kubevirt will not be able to correlated between the desired (or domain) interface and the one seen by the guest-agent.
This is true today and that will remain so with this enhancement.

We may or may not be interested in addressing this in the future. However, I would expect that we always reflect the permanent MAC, like today.

RamLavi commented:
PR that sets the SR-IOV on the vmi.status according to [said](https://github.com/kubevirt/enhancements/pull/149/changes#diff-185f8ffa8d5c01c22c8a40cceb5659d810167596ac8dfb678ecf91d628a7aec9R112-R114):  https://github.com/kubevirt/kubevirt/pull/16485

0xFelix reviewed (commented):
Did you think about GitOps? Ideally we should never touch the spec of an object. Can you not use the status to temporarily persist the MAC for the lifetime of the VM object?
